File: SignalUI/Views/OWSSearchBar.m
//
// Copyright 2018 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "OWSSearchBar.h"
#import "UIView+SignalUI.h"
#import <SignalUI/SignalUI-Swift.h>

NS_ASSUME_NONNULL_BEGIN

@interface OWSSearchBar ()

@property (nonatomic) OWSSearchBarStyle currentStyle;

@end

@implementation OWSSearchBar

@synthesize searchFieldBackgroundColorOverride = _searchFieldBackgroundColorOverride;

- (instancetype)init
{
    if (self = [super init]) {
        [self ows_configure];
    }

    return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    if (self = [super initWithFrame:frame]) {
        [self ows_configure];
    }

    return self;
}

- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder
{
    if (self = [super initWithCoder:aDecoder]) {
        [self ows_configure];
    }

    return self;
}

- (void)ows_configure
{
    _currentStyle = OWSSearchBarStyle_Default;

    [self ows_applyTheme];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(themeDidChange:)
                                                 name:NSNotification.ThemeDidChange
                                               object:nil];
}

- (void)ows_applyTheme
{
    [self.class applyThemeToSearchBar:self style:self.currentStyle];
}

+ (void)applyThemeToSearchBar:(UISearchBar *)searchBar
{
    [self applyThemeToSearchBar:searchBar style:OWSSearchBarStyle_Default];
}

+ (void)applyThemeToSearchBar:(UISearchBar *)searchBar style:(OWSSearchBarStyle)style
{
    OWSAssertIsOnMainThread();

    UIColor *foregroundColor = Theme.secondaryTextAndIconColor;
    searchBar.tintColor = Theme.secondaryTextAndIconColor;
    searchBar.barStyle = Theme.barStyle;
    searchBar.barTintColor = Theme.backgroundColor;

    // Hide searchBar border.
    // Alternatively we could hide the border by using `UISearchBarStyleMinimal`, but that causes an issue when toggling
    // from light -> dark -> light theme wherein the textField background color appears darker than it should
    // (regardless of our re-setting textfield.backgroundColor below).
    searchBar.backgroundImage = [UIImage new];

    if (Theme.isDarkThemeEnabled) {
        UIImage *clearImage = [UIImage imageNamed:@"searchbar_clear"];
        [searchBar setImage:[clearImage asTintedImageWithColor:foregroundColor]
            forSearchBarIcon:UISearchBarIconClear
                       state:UIControlStateNormal];

        UIImage *searchImage = [UIImage imageNamed:@"searchbar_search"];
        [searchBar setImage:[searchImage asTintedImageWithColor:foregroundColor]
            forSearchBarIcon:UISearchBarIconSearch
                       state:UIControlStateNormal];
    } else {
        [searchBar setImage:nil forSearchBarIcon:UISearchBarIconClear state:UIControlStateNormal];

        [searchBar setImage:nil forSearchBarIcon:UISearchBarIconSearch state:UIControlStateNormal];
    }

    UIColor *searchFieldBackgroundColor = Theme.searchFieldBackgroundColor;
    if ([searchBar isKindOfClass:[OWSSearchBar class]]
        && ((OWSSearchBar *)searchBar).searchFieldBackgroundColorOverride) {
        searchFieldBackgroundColor = ((OWSSearchBar *)searchBar).searchFieldBackgroundColorOverride;
    }

    [searchBar traverseViewHierarchyDownwardWithVisitor:^(UIView *view) {
        if ([view isKindOfClass:[UITextField class]]) {
            UITextField *textField = (UITextField *)view;
            textField.backgroundColor = searchFieldBackgroundColor;
            textField.textColor = Theme.primaryTextColor;
            textField.keyboardAppearance = Theme.keyboardAppearance;
        }
    }];
}

- (void)switchToStyle:(OWSSearchBarStyle)style
{
    self.currentStyle = style;
    [self ows_applyTheme];
}

- (void)themeDidChange:(NSNotification *)notification
{
    OWSAssertIsOnMainThread();

    [self ows_applyTheme];
}

- (void)setSearchFieldBackgroundColorOverride:(nullable UIColor *)searchFieldBackgroundColorOverride
{
    OWSAssertIsOnMainThread();

    _searchFieldBackgroundColorOverride = searchFieldBackgroundColorOverride;

    [self ows_applyTheme];
}

- (nullable UIColor *)searchFieldBackgroundColorOverride
{
    OWSAssertIsOnMainThread();
    return _searchFieldBackgroundColorOverride;
}

@end

NS_ASSUME_NONNULL_END

File: SignalUI/Views/OWSSearchBar.swift
//
// Copyright 2023 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import SignalCoreKit

public class OWSSearchBar: UISearchBar {

    public override init(frame: CGRect) {
        super.init(frame: frame)
        configure()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        configure()
    }

    // MARK: -

    public var searchFieldBackgroundColorOverride: UIColor? {
        didSet {
            applyTheme()
        }
    }

    private func configure() {
        applyTheme()

        NotificationCenter.default.addObserver(self, selector: #selector(themeDidChange), name: .themeDidChange, object: nil)
    }

    // MARK: Theme

    public static func applyTheme(to searchBar: UISearchBar) {
        AssertIsOnMainThread()

        searchBar.tintColor = Theme.secondaryTextAndIconColor
        searchBar.barStyle = Theme.barStyle
        searchBar.barTintColor = Theme.backgroundColor

        // Hide searchBar border.
        // Alternatively we could hide the border by using `UISearchBarStyleMinimal`, but that causes an issue when toggling
        // from light -> dark -> light theme wherein the textField background color appears darker than it should
        // (regardless of our re-setting textfield.backgroundColor below).
        searchBar.backgroundImage = UIImage()

        if Theme.isDarkThemeEnabled {
            let foregroundColor = Theme.secondaryTextAndIconColor

            let clearImage = UIImage(imageLiteralResourceName: "x-circle-fill")
            searchBar.setImage(
                clearImage.asTintedImage(color: foregroundColor),
                for: .clear,
                state: .normal
            )

            let searchImage = UIImage(imageLiteralResourceName: "search")
            searchBar.setImage(
                searchImage.asTintedImage(color: foregroundColor),
                for: .search,
                state: .normal
            )
        } else {
            searchBar.setImage(nil, for: .clear, state: .normal)
            searchBar.setImage(nil, for: .search, state: .normal)
        }

        let searchFieldBackgroundColor: UIColor
        if let owsSearchBar = searchBar as? OWSSearchBar, let colorOverride = owsSearchBar.searchFieldBackgroundColorOverride {
            searchFieldBackgroundColor = colorOverride
        } else {
            searchFieldBackgroundColor = Theme.searchFieldBackgroundColor
        }

        searchBar.traverseHierarchyDownward { view in
            guard let textField = view as? UITextField else { return }
            textField.backgroundColor = searchFieldBackgroundColor
            textField.textColor = Theme.primaryTextColor
            textField.keyboardAppearance = Theme.keyboardAppearance
        }
    }

    private func applyTheme() {
        Self.applyTheme(to: self)
    }

    @objc
    private func themeDidChange(_ notification: Notification) {
        AssertIsOnMainThread()
        applyTheme()
    }
}
